src/qibolab/result.py

```diff
@@ -1,10 +1,168 @@
+    def serialize(self):
+        """Serialize as a dictionary."""
+        serialized_dict = {
+            "MSR[V]": self.magnitude.flatten(),
+            "i[V]": self.voltage_i.flatten(),
+            "q[V]": self.voltage_q.flatten(),
+            "phase[rad]": self.phase.flatten(),
+        }
+        return serialized_dict
+
+    @property
+    def average(self):
+        """Perform average over i and q"""
+        average_data = np.mean(self.voltage, axis=0)
+        std_data = np.std(self.voltage, axis=0, ddof=1) / np.sqrt(self.voltage.shape[0])
+        return AveragedIntegratedResults(average_data, std_data)
+
+
+class AveragedIntegratedResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.INTEGRATION and AveragingMode.CYCLIC
+    or the averages of ``IntegratedResults``
+    """
+
+    def __init__(self, data: np.ndarray, std: np.ndarray = np.array([])):
+        super().__init__(data)
+        self.std: Optional[npt.NDArray[np.float64]] = std
+
+    def __add__(self, data):
+        new_res = super().__add__(data)
+        new_res.std = np.append(self.std, data.std)
+        return new_res
+
+
+class RawWaveformResults(IntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.SINGLESHOT
+    may also be used to store the integration weights ?
+    """
+
+
+class AveragedRawWaveformResults(AveragedIntegratedResults):
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.RAW and AveragingMode.CYCLIC
+    or the averages of ``RawWaveformResults``
+    """
+
+
+class StateResults:
+    """
+    Data structure to deal with the output of
+    :func:`qibolab.platforms.abstract.AbstractPlatform.execute_pulse_sequence`
+    :func:`qibolab.platforms.abstract.AbstractPlatform.sweep`
+
+    Associated with AcquisitionType.DISCRIMINATION and AveragingMode.SINGLESHOT
+    """
+
+    def __init__(self, data: np.ndarray):
+        self.samples: npt.NDArray[np.uint32] = data
+
+    def __add__(self, data):
+        return self.__class__(np.append(self.samples, data.samples))
+
+    @lru_cache
+    def probability(self, state=0):
+        """Returns the statistical frequency of the specified state (0 or 1)."""
+        return abs(1 - state - np.mean(self.samples, axis=0))
+
+    @property
+    def serialize(self):
+        """Serialize as a dictionary."""
+        serialized_dict = {
+            "state_0": self.probability(0).flatten(),
```

Comment:
Since this won't even be a `StateResults`, maybe this can be simplified to a more agnostic:
```suggestion
            0: self.probability(0).flatten(),
```

Reply:
Yeah better, I would avoid naming things as far as I can